api/lib/validate.js

```diff
@@ -166,6 +174,177 @@ function validateOrderbook(orderbook) {
+        'Invalid parameter: MemoType. MemoType must be a string');
+    }
+    if (!/(undefined|string)/.test(typeof memo.MemoData)) {
+      throw error(
+        'Invalid parameter: MemoData. MemoData must be a string');
+    }
+    if (!memo.MemoData && !memo.MemoType) {
+      throw error('Missing parameter: '
+        + 'MemoData or MemoType. For a memo object MemoType or MemoData '
+        + 'are both optional, as long as one of them is present');
+    }
+  }
+}
+
+function validatePayment(payment) {
+  if (!isValidAddress(payment.source_account)) {
+    throw error('Invalid parameter: source_account. '
+      + 'Must be a valid Ripple address');
+  }
+
+  if (!isValidAddress(payment.destination_account)) {
+    throw error('Invalid parameter: '
+      + 'destination_account. Must be a valid Ripple address');
+  }
+
+  if (payment.source_tag &&
+      (!validator.isValid(payment.source_tag, 'UINT32'))) {
+    throw error('Invalid parameter: source_tag. '
+      + 'Must be a string representation of an unsiged 32-bit integer');
+  }
+
+  if (payment.destination_tag
+      && (!validator.isValid(payment.destination_tag, 'UINT32'))) {
+    throw error('Invalid parameter: '
+      + 'destination_tag. Must be a string representation of an unsiged '
+      + '32-bit integer');
+  }
+
+  if (!payment.destination_amount
+      || (!validator.isValid(payment.destination_amount, 'Amount'))) {
+    throw error('Invalid parameter: '
+      + 'destination_amount. Must be a valid Amount object');
+  }
+
+  if (payment.source_amount   // source_amount is optional
+      && (!validator.isValid(payment.source_amount, 'Amount'))) {
+    throw error(
+      'Invalid parameter: source_amount. Must be a valid Amount object');
+  }
+
+  if (payment.destination_amount
+      && payment.destination_amount.currency.toUpperCase() === 'XRP'
+      && payment.destination_amount.issuer) {
+    throw error(
+      'Invalid parameter: destination_amount. XRP cannot have issuer');
+  }
+  if (payment.source_amount
+      && payment.source_amount.currency.toUpperCase() === 'XRP'
+      && payment.source_amount.issuer) {
+    throw error(
+      'Invalid parameter: source_amount. XRP cannot have issuer');
+  }
+
+  if (payment.source_slippage
+      && !validator.isValid(payment.source_slippage, 'FloatString')) {
+    throw error(
+      'Invalid parameter: source_slippage. Must be a valid FloatString');
+  }
+
+  if (payment.invoice_id
+      && !validator.isValid(payment.invoice_id, 'Hash256')) {
+    throw error(
+      'Invalid parameter: invoice_id. Must be a valid Hash256');
+  }
+
+  if (payment.paths) {
+    if (typeof payment.paths === 'string') {
+      try {
+        JSON.parse(payment.paths);
+      } catch (exception) {
+        throw error(
+          'Invalid parameter: paths. Must be a valid JSON string or object');
+      }
+    } else if (typeof payment.paths === 'object') {
+      try {
+        JSON.parse(JSON.stringify(payment.paths));
```

Comment:
Yes, for example cyclic data structures (http://www.json.org/js.html) or objects that define a toJSON method that can throw an exception.